// testdouble/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const function_1 = require("./function");
const object_1 = require("./object");
const constructor_1 = require("./constructor");
const instance_1 = require("./instance");
const imitate_1 = require("./imitate");
const when_1 = require("./when");
const verify_1 = require("./verify");
const matchers_1 = require("./matchers");
const replace_1 = require("./replace");
const explain_1 = require("./explain");
const reset_1 = require("./reset");
const config_1 = require("./config");
const callback_1 = require("./callback");
const version_1 = require("./version");
const quibble = require("quibble");
module.exports = {
    function: function_1.default,
    func: function_1.default,
    object: object_1.default,
    constructor: constructor_1.default,
    instance: instance_1.default,
    imitate: imitate_1.default,
    when: when_1.default,
    verify: verify_1.default,
    matchers: matchers_1.default,
    replace: replace_1.default,
    replaceEsm: replace_1.replaceEsm,
    explain: explain_1.default,
    reset: reset_1.default,
    config: config_1.default,
    callback: callback_1.default,
    version: version_1.default,
    quibble
};
